src/main/scala/org/apache/spark/shuffle/ucx/NvkvShuffleMapOutputWriter.scala

```diff
@@ -0,0 +1,291 @@
+  private var totalPartitionsPadding: Long = 0
+  final private var bufferSize = 4096
+  private var lastPartitionId = -1
+  private var currChannelPosition = 0L
+  private var dsIdx: Int = 0
+  private var bytesWrittenToMergedFile = 0L
+  private var outputBufferedFileStream: BufferedOutputStream = null
+  private var nvkvWrapper: NvkvWrapper = ucxTransport.getNvkvWrapper
+  private var executerId = SparkEnv.get.blockManager.blockManagerId.executorId.toLong
+
+  private def getBlockOffset = {
+    //TODO - add executer partition offset
+    val numShuffles = 1
+    val numOfMappers = SparkEnv.get.conf.getInt("spark.groupByTest.numMappers", 1)
+    val shuffleBlockSize = nvkvWrapper.getPartitionSize / numShuffles
+    val mapBlockSize = shuffleBlockSize / numOfMappers
+    val r = mapBlockSize % 512;
+    val alignedMapBlockSize = if (r != 0) {mapBlockSize - (r)} else mapBlockSize
+    (shuffleId * shuffleBlockSize) + (mapId * alignedMapBlockSize)
+  }
+
+  @throws[IOException]
+  override def getPartitionWriter(reducePartitionId: Int): ShufflePartitionWriter = {
+    NvkvShuffleMapOutputWriter.log.debug(s"NvkvShuffleMapOutputWriter getPartitionWriter for reducePartition $reducePartitionId")
+    if (reducePartitionId <= lastPartitionId) throw new IllegalArgumentException("Partitions should be requested in increasing order.")
+    lastPartitionId = reducePartitionId
+    currChannelPosition = getBlockOffset + bytesWrittenToMergedFile + totalPartitionsPadding
+    /* 
+     * Temporary solution for distributing the load between the NVME devices to get full read BW.
+     * This solution is not optimal because is nt taking into account NUMA topology and it's dependent in randmoness.
+     * See SparkDPU documentation for more information. 
+     */
+    dsIdx = Random.nextInt(nvkvWrapper.getNumOfDevices)
+    NvkvShuffleMapOutputWriter.log.debug(s"NvkvShuffleMapOutputWriter reducePartition offset $currChannelPosition")
+    new NvkvShufflePartitionWriter(reducePartitionId)
+  }
+
+  @throws[IOException]
+  override def commitAllPartitions: Array[Long] = {
+    NvkvShuffleMapOutputWriter.log.debug("NvkvShuffleMapOutputWriter commitAllPartitions")
+    // Check the position after transferTo loop to see if it is in the right position and raise a
+    // exception if it is incorrect. The position will not be increased to the expected length
+    // after calling transferTo in kernel version 2.6.32. This issue is described at
+    // https://bugs.openjdk.java.net/browse/JDK-7052359 and SPARK-3948.
+
+    var blockOffset = getBlockOffset;
+
+    //TODO - move to shuffleTransport    
+    var packMapperData: ByteBuffer = ByteBuffer.allocateDirect(UnsafeUtils.INT_SIZE + // numOfMappers
+                                                               UnsafeUtils.INT_SIZE + // numOfReducePartitions
+                                                               UnsafeUtils.INT_SIZE + // mapperId
```

Comment:
Maybe just `3 * UnsafeUtils.INT_SIZE` and comment `// mappers + reduce partitions + id` above it?

Reply:
Writing above it when all the next are in the same line seems strange.
IMO this syntax is more clear, we know exactly how much each field takes.